flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala

```diff
@@ -0,0 +1,468 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.table.typeinfo
+
+import java.util
+
+import org.apache.flink.api.common.typeutils.base.BasicTypeComparator
+import org.apache.flink.api.common.typeutils.{CompositeTypeComparator, TypeComparator, TypeSerializer}
+import org.apache.flink.api.java.typeutils.runtime.TupleComparatorBase
+import org.apache.flink.api.table.Row
+import org.apache.flink.api.table.typeinfo.NullMaskUtils.readIntoNullMask
+import org.apache.flink.api.table.typeinfo.RowComparator.NullChecker
+import org.apache.flink.core.memory.{MemorySegment, DataOutputView, DataInputView}
+import org.apache.flink.types.{KeyFieldOutOfBoundsException, NullKeyFieldException}
+
+/**
+ * Comparator for [[Row]].
+ */
+class RowComparator private (
+    /** key positions describe which fields are keys in what order */
+    val keyPositions: Array[Int],
+    /** comparators for the key fields, in the same order as the key fields */
+    val comparators: Array[TypeComparator[_]],
+    /** serializers to deserialize the first n fields for comparison */
+    val serializers: Array[TypeSerializer[_]],
+    /** auxiliary fields for normalized key support */
+    private var normalizedKeyLengths: Array[Int],
+    private var numLeadingNormalizableKeys: Int,
+    private var normalizableKeyPrefixLen: Int,
+    private var invertNormKey: Boolean)
+  extends CompositeTypeComparator[Row] with Serializable {
+
+  // null masks for serialized comparison
+  private val nullMask1 = new Array[Boolean](serializers.length)
+  private val nullMask2 = new Array[Boolean](serializers.length)
+
+  // cache for the deserialized key field objects
+  @transient
+  private var deserializedKeyFields1: Array[Any] = null
+  @transient
+  private var deserializedKeyFields2: Array[Any] = null
+
+  // null checker for reference comparison
+  private val nullChecker = new NullChecker()
+
+  def this(
+      keyPositions: Array[Int],
+      comparators: Array[TypeComparator[_]],
+      serializers: Array[TypeSerializer[_]]) = {
+    this(keyPositions, comparators, serializers, new Array[Int](keyPositions.length), 0, 0, false)
+    // set up auxiliary fields for normalized key support
+    setupAuxiliaryFields()
```

Comment:
Can't we simply let `setupAuxiliaryFields` return the `normalizedKeyLengths`, `numLeadingNormalizableKeys` and `normalizableKeyPrefixLen` values for the constructor call? Maybe we should then rename the method to `createAuxiliaryFields`.